fields_error.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"go.uber.org/zap/zapcore"
+)
+
+type errorWithFields struct {
+	err    error
+	fields []Field
+}
+
+// WrapError returns an error that will log the provided fields if the error
+// is logged using `Error`.
+func WrapError(err error, fields ...Field) error {
+	return &errorWithFields{
+		err:    err,
+		fields: fields,
+	}
+}
+
+func (e errorWithFields) Unwrap() error {
+	return e.err
+}
+
+func (e errorWithFields) Error() string {
+	return e.err.Error()
+}
+
+func (e errorWithFields) MarshalLogObject(oe zapcore.ObjectEncoder) error {
```

Comment:
Using the same interface as `ObjectMarshaler` makes it easy to integrate in an efficient way. I'm open to using a separate interface, but then adding it as a nested object will require an additional allocation to wrap the custom interface into an `ObjectMarshaler`

Reply:
I think it's fine to use implement `ObjectMarshaler` directly, yeah.

That said, I think it might be nicer to do

```go
func WrappedError(err error, fields ...Field) Field {
  return zap.Error(&errorWithFields{
    err:    err,
    fields: fields,
  })
}
``` 

primarily for ergonomics, because (1) that's what you're going to be doing with it anyway and (2) it probably makes more sense as a `Field` func rather than as something you pass to a `Field` func (since the only practical candidate is `Error`).

(pedantry: I chose `WrappedError` because none of the other `Field` func names are/have verbs)

Reply:
I think we want the return to be an `error` to match the general usage of this -- the method that is adding fields currently returns an error, and the signature shouldn't change, this just adds some fields to be logged as part of the error log.

I've added an example to show this that should showcase it a little more clearly.